File: studypilot_backend/apps/accounts/sui.py
"""Verify Sui wallet personal-message signatures.

A wallet proves ownership of an address by signing a challenge with
`signPersonalMessage`. Sui does not sign the raw bytes: it prefixes an intent,
BCS-encodes the message, hashes with blake2b-256, and signs that digest. To
accept a login we have to reproduce exactly the same digest, check the
signature against the supplied public key, and then confirm that public key
actually derives the address the client claims. Skipping that last step would
let anyone log in as anyone by sending their own valid signature alongside
someone else's address.

Ed25519 only. Sui also allows Secp256k1 and Secp256r1 keys, which are rejected
here rather than silently mis-verified.
"""
import base64
import hashlib
import re

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

SIGNATURE_SCHEME_ED25519 = 0x00
ED25519_SIGNATURE_LENGTH = 64
ED25519_PUBLIC_KEY_LENGTH = 32
# IntentScope::PersonalMessage, IntentVersion::V0, AppId::Sui
PERSONAL_MESSAGE_INTENT = bytes([3, 0, 0])
SUI_ADDRESS_RE = re.compile(r"^0x[0-9a-fA-F]{64}$")


class SuiVerificationError(Exception):
    pass


def _uleb128(value):
    """BCS encodes a vector length as ULEB128."""
    out = bytearray()
    while True:
        byte = value & 0x7F
        value >>= 7
        if value:
            out.append(byte | 0x80)
        else:
            out.append(byte)
            return bytes(out)


def normalize_address(address):
    """Lowercase and zero-pad to the canonical 32-byte hex form."""
    if not address or not isinstance(address, str):
        raise SuiVerificationError("A Sui address is required.")
    value = address.strip().lower()
    if not value.startswith("0x"):
        raise SuiVerificationError("Sui address must start with 0x.")
    body = value[2:]
    if not body or len(body) > 64 or not all(c in "0123456789abcdef" for c in body):
        raise SuiVerificationError("Sui address is not valid hex.")
    return "0x" + body.rjust(64, "0")


def address_from_public_key(public_key_bytes, scheme=SIGNATURE_SCHEME_ED25519):
    """Sui address = blake2b-256(flag || public_key)."""
    digest = hashlib.blake2b(bytes([scheme]) + public_key_bytes, digest_size=32).digest()
    return "0x" + digest.hex()


def personal_message_digest(message_bytes):
    """The digest a Sui wallet actually signs for a personal message."""
    bcs_message = _uleb128(len(message_bytes)) + message_bytes
    return hashlib.blake2b(PERSONAL_MESSAGE_INTENT + bcs_message, digest_size=32).digest()


def verify_personal_message(message, signature_b64, claimed_address):
    """Return the verified Sui address, or raise SuiVerificationError.

    `signature_b64` is the wallet's serialized signature:
    flag(1) || signature(64) || public_key(32), base64 encoded.
    """
    if not signature_b64 or not isinstance(signature_b64, str):
        raise SuiVerificationError("A wallet signature is required.")

    try:
        raw = base64.b64decode(signature_b64, validate=True)
    except Exception as exc:
        raise SuiVerificationError("Signature is not valid base64.") from exc

    expected = 1 + ED25519_SIGNATURE_LENGTH + ED25519_PUBLIC_KEY_LENGTH
    if len(raw) != expected:
        raise SuiVerificationError("Unexpected signature length.")

    scheme = raw[0]
    if scheme != SIGNATURE_SCHEME_ED25519:
        raise SuiVerificationError("Only Ed25519 wallet signatures are supported.")

    signature = raw[1 : 1 + ED25519_SIGNATURE_LENGTH]
    public_key_bytes = raw[1 + ED25519_SIGNATURE_LENGTH :]

    digest = personal_message_digest(message.encode("utf-8") if isinstance(message, str) else message)
    try:
        Ed25519PublicKey.from_public_bytes(public_key_bytes).verify(signature, digest)
    except InvalidSignature as exc:
        raise SuiVerificationError("Wallet signature could not be verified.") from exc
    except Exception as exc:
        raise SuiVerificationError("Wallet public key is not valid.") from exc

    derived = address_from_public_key(public_key_bytes, scheme)
    # The signature alone only proves control of *some* key. Without this the
    # caller could pair a real signature with somebody else's address.
    if derived != normalize_address(claimed_address):
        raise SuiVerificationError("Signature does not match the given Sui address.")
    return derived

File: studypilot_worker/src/auth/sui.ts
// Verify Sui wallet personal-message signatures. Port of apps/accounts/sui.py.
//
// Sui does not sign the raw bytes: it prefixes an intent, BCS-encodes the
// message, hashes with blake2b-256, and signs that digest. We reproduce the
// digest, check the signature against the supplied public key, then confirm
// that key derives the claimed address. Skipping the last step would let anyone
// log in as anyone by pairing their own valid signature with another address.
//
// Ed25519 only; Secp256k1/r1 keys are rejected rather than mis-verified.

import { blake2b } from "@noble/hashes/blake2.js";

const SIGNATURE_SCHEME_ED25519 = 0x00;
const ED25519_SIGNATURE_LENGTH = 64;
const ED25519_PUBLIC_KEY_LENGTH = 32;
// IntentScope::PersonalMessage, IntentVersion::V0, AppId::Sui
const PERSONAL_MESSAGE_INTENT = [3, 0, 0];

export class SuiVerificationError extends Error {}

function uleb128(value: number): number[] {
  const out: number[] = [];
  for (;;) {
    const byte = value & 0x7f;
    value >>>= 7;
    if (value) out.push(byte | 0x80);
    else {
      out.push(byte);
      return out;
    }
  }
}

const hex = (bytes: Uint8Array) => Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");

export function normalizeAddress(address: unknown): string {
  if (!address || typeof address !== "string") throw new SuiVerificationError("A Sui address is required.");
  const value = address.trim().toLowerCase();
  if (!value.startsWith("0x")) throw new SuiVerificationError("Sui address must start with 0x.");
  const body = value.slice(2);
  if (!body || body.length > 64 || !/^[0-9a-f]+$/.test(body)) throw new SuiVerificationError("Sui address is not valid hex.");
  return "0x" + body.padStart(64, "0");
}

export function addressFromPublicKey(publicKey: Uint8Array, scheme = SIGNATURE_SCHEME_ED25519): string {
  return "0x" + hex(blake2b(new Uint8Array([scheme, ...publicKey]), { dkLen: 32 }));
}

export function personalMessageDigest(message: Uint8Array): Uint8Array<ArrayBuffer> {
  return blake2b(new Uint8Array([...PERSONAL_MESSAGE_INTENT, ...uleb128(message.length), ...message]), { dkLen: 32 });
}

function decodeBase64Strict(value: string): Uint8Array {
  if (!/^[A-Za-z0-9+/]*={0,2}$/.test(value) || value.length % 4 !== 0) {
    throw new SuiVerificationError("Signature is not valid base64.");
  }
  try {
    return Uint8Array.from(atob(value), (c) => c.charCodeAt(0));
  } catch {
    throw new SuiVerificationError("Signature is not valid base64.");
  }
}

/** Return the verified address, or throw SuiVerificationError. */
export async function verifyPersonalMessage(message: string, signatureB64: string, claimedAddress: string): Promise<string> {
  if (!signatureB64 || typeof signatureB64 !== "string") throw new SuiVerificationError("A wallet signature is required.");
  const raw = decodeBase64Strict(signatureB64);
  if (raw.length !== 1 + ED25519_SIGNATURE_LENGTH + ED25519_PUBLIC_KEY_LENGTH) {
    throw new SuiVerificationError("Unexpected signature length.");
  }
  if (raw[0] !== SIGNATURE_SCHEME_ED25519) throw new SuiVerificationError("Only Ed25519 wallet signatures are supported.");
  const signature = raw.slice(1, 1 + ED25519_SIGNATURE_LENGTH);
  const publicKey = raw.slice(1 + ED25519_SIGNATURE_LENGTH);
  const digest = personalMessageDigest(new TextEncoder().encode(message));

  let valid = false;
  try {
    // Native Ed25519 in workerd: far cheaper than a JS implementation.
    const key = await crypto.subtle.importKey("raw", publicKey, { name: "Ed25519" }, false, ["verify"]);
    valid = await crypto.subtle.verify({ name: "Ed25519" }, key, signature, digest);
  } catch {
    throw new SuiVerificationError("Wallet public key is not valid.");
  }
  if (!valid) throw new SuiVerificationError("Wallet signature could not be verified.");

  const derived = addressFromPublicKey(publicKey);
  if (derived !== normalizeAddress(claimedAddress)) {
    throw new SuiVerificationError("Signature does not match the given Sui address.");
  }
  return derived;
}
